### PR TITLE
Reduce HttpWebResponse header allocations

### PR DESCRIPTION
**Pull request by bbowyersmyth**
Removed unnecessary ToUpperInvariant allocation in WinHttpHandler.

HttpWebResponse created a StringBuilder for every header in the response to handle possible multi-values. Common case is a single value so lazy create the buffer when there is more than 1.

### CODE REVIEW

**Comment by stephentoub** [comment]
LGTM

cc: @davidsh, @CIPop, @SidharthNabar

File: src/System.Net.Requests/src/System/Net/HttpWebResponse.cs [hunk 1]
```diff
@@ -167,22 +167,29 @@ private void CheckDisposed()
...
-            while (enumerator.MoveNext())
+            string headerValue = enumerator.Current;
+
+            if (enumerator.MoveNext())
             {
-                buffer.Append(", ");
-                buffer.Append(enumerator.Current);
+                // Multi-valued header
+                var buffer = new StringBuilder();
+                buffer.Append(headerValue);
```
Comment (Maxwe11) [inline src/System.Net.Requests/src/System/Net/HttpWebResponse.cs]: :bulb:  `var buffer = new StringBuilder(headerValue);`

File: src/System.Net.Requests/src/System/Net/HttpWebResponse.cs [hunk 2]
```diff
@@ -167,22 +167,29 @@ private void CheckDisposed()
...
+
+                do
+                {
+                    buffer.Append(", ");
+                    buffer.Append(enumerator.Current);
+                } while (enumerator.MoveNext());
```
Comment (jasonwilliams200OK) [inline src/System.Net.Requests/src/System/Net/HttpWebResponse.cs]: can be reduced to one line:

```c#
do
{
    buffer.Append(", ").Append(enumerator.Current);
} while (enumerator.MoveNext());
```

Comment (davidsh) [inline src/System.Net.Requests/src/System/Net/HttpWebResponse.cs]: That doesn't really save any performance and it's harder to debug those kinds of constructs.  It is better to leave it as-is on 2 lines.

File: src/System.Net.Requests/src/System/Net/HttpWebResponse.cs (hunk 1 above)
Comment (bbowyersmyth) [inline src/System.Net.Requests/src/System/Net/HttpWebResponse.cs]: Changed to use StringBuilder(string)

**Comment by nguerrera** [comment]
@dotnet-bot  test this please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@davidsh, does this look ok to you?

**Comment by davidsh** [comment]
@stephentoub We need to run extra tests on this that aren't in GitHub.

**Comment by davidsh** [comment]
Also, as I mentioned above, this construct should be split into 2 lines:
` buffer.Append(", ").Append(enumerator.Current);`

**Comment by stephentoub** [comment]
> We need to run extra tests on this that aren't in GitHub

Ok. Thanks for adding the associated tag.

**Comment by davidsh** [comment]
LGTM

**Pull request merged**